File: binary-tree-inorder-traversal.cpp
class Solution {
public:
    vector<int> inorderTraversal(TreeNode* root) {
        vector<int> ans;
        stack<TreeNode*> stk;
        TreeNode* cur = root;
        while(cur || !stk.empty()){
            //if cur is nullptr, it's ok, because it will later be stk.top()
            while(cur){
                stk.push(cur);
                cur = cur->left;
            }
            cur = stk.top(); stk.pop();
            ans.push_back(cur->val);
            //it's fine if cur->right is nullptr
            cur = cur->right;
        }
        
        return ans;
    }
};

File: binary-tree-inorder-traversal.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[]}
 */
var inorderTraversal = function(root) {
    const node = [];
    if(!root) return node;
    node.push(...inorderTraversal(root.left));
    node.push(root.val);
    node.push(...inorderTraversal(root.right));
    return node;
};
